Name: Muley, Tushar

Assignment: DSC 540 Term Project

# Milestone 4: Cleaning/Formatting API Source
Week 9 and 10

In [51]:
#Import Libraries
import urllib3
from urllib3 import request
import json
import pandas as pd
import numpy as np
import requests

In [48]:
#Working code for pulling race results for 2020

r = requests.get('http://ergast.com/api/f1/2020/results/1.json') #API for race results
r = r.json()
races = r['MRData']['RaceTable']['Races'] #One list above results
#r['MRData']['RaceTable']['Races'][0]['Results'] #On results
#circuit =  r["MRData"]["RaceTable"]["Races"]["Circuit"] #Circuits but fails
#Print what I find
races 

[{'season': '2020',
  'round': '1',
  'url': 'https://en.wikipedia.org/wiki/2020_Austrian_Grand_Prix',
  'raceName': 'Austrian Grand Prix',
  'Circuit': {'circuitId': 'red_bull_ring',
   'url': 'http://en.wikipedia.org/wiki/Red_Bull_Ring',
   'circuitName': 'Red Bull Ring',
   'Location': {'lat': '47.2197',
    'long': '14.7647',
    'locality': 'Spielburg',
    'country': 'Austria'}},
  'date': '2020-07-05',
  'time': '13:10:00Z',
  'Results': [{'number': '77',
    'position': '1',
    'positionText': '1',
    'points': '25',
    'Driver': {'driverId': 'bottas',
     'permanentNumber': '77',
     'code': 'BOT',
     'url': 'http://en.wikipedia.org/wiki/Valtteri_Bottas',
     'givenName': 'Valtteri',
     'familyName': 'Bottas',
     'dateOfBirth': '1989-08-28',
     'nationality': 'Finnish'},
    'Constructor': {'constructorId': 'mercedes',
     'url': 'http://en.wikipedia.org/wiki/Mercedes-Benz_in_Formula_One',
     'name': 'Mercedes',
     'nationality': 'German'},
    'grid': '1',


In [49]:
#Race Results working version
r = requests.get('http://ergast.com/api/f1/2020/results/1.json') #API for race results
r = r.json()
results = r['MRData']['RaceTable']['Races'][0]['Results']

In [50]:
results

[{'number': '77',
  'position': '1',
  'positionText': '1',
  'points': '25',
  'Driver': {'driverId': 'bottas',
   'permanentNumber': '77',
   'code': 'BOT',
   'url': 'http://en.wikipedia.org/wiki/Valtteri_Bottas',
   'givenName': 'Valtteri',
   'familyName': 'Bottas',
   'dateOfBirth': '1989-08-28',
   'nationality': 'Finnish'},
  'Constructor': {'constructorId': 'mercedes',
   'url': 'http://en.wikipedia.org/wiki/Mercedes-Benz_in_Formula_One',
   'name': 'Mercedes',
   'nationality': 'German'},
  'grid': '1',
  'laps': '71',
  'status': 'Finished',
  'Time': {'millis': '5455739', 'time': '1:30:55.739'},
  'FastestLap': {'rank': '2',
   'lap': '68',
   'Time': {'time': '1:07.657'},
   'AverageSpeed': {'units': 'kph', 'speed': '229.758'}}}]

In [21]:
#Pull data from API for races

#My dictionary with empty lists for inputs
races = {'season': [],
        'round': [],
        'circuit_id': [],
        'lat': [],
        'long': [],
        'country': [],
        'date': [],
        'url': []}

for year in list(range(1950,2020)): #Loop for year range
    
    url = 'https://ergast.com/api/f1/{}.json' #Race URL
    r = requests.get(url.format(year)) #Request using year
    json = r.json() #Read the returning JSON

#Loop through the JSON appending data that matches list
    for item in json['MRData']['RaceTable']['Races']:
            races['season'].append(int(item['season']))
            races['round'].append(int(item['round']))
            races['circuit_id'].append(item['Circuit']['circuitId'])
            races['lat'].append(float(item['Circuit']['Location']['lat']))
            races['long'].append(float(item['Circuit']['Location']['long']))
            races['country'].append(item['Circuit']['Location']['country'])
            races['date'].append(item['date'])
            races['url'].append(item['url'])
#Take the list and make it into a dataframe
df_races = pd.DataFrame(races)

In [22]:
#Review the data in races
df_races

season  round    circuit_id      lat       long      country  \
0       1950      1   silverstone  52.0786   -1.01694           UK   
1       1950      2        monaco  43.7347    7.42056       Monaco   
2       1950      3  indianapolis  39.7950  -86.23470          USA   
3       1950      4    bremgarten  46.9589    7.40194  Switzerland   
4       1950      5           spa  50.4372    5.97139      Belgium   
...      ...    ...           ...      ...        ...          ...   
1013    2019     17        suzuka  34.8431  136.54100        Japan   
1014    2019     18     rodriguez  19.4042  -99.09070       Mexico   
1015    2019     19      americas  30.1328  -97.64110          USA   
1016    2019     20    interlagos -23.7036  -46.69970       Brazil   
1017    2019     21    yas_marina  24.4672   54.60310          UAE   

            date                                                url  
0     1950-05-13  http://en.wikipedia.org/wiki/1950_British_Gran...  
1     1950-05-21  http://en.wikipedia.org/wiki/1950_Monaco_Grand...  
2     1950-05-30  http://en.wikipedia.org/wiki/1950_Indianapolis...  
3     1950-06-04  http://en.wikipedia.org/wiki/1950_Swiss_Grand_...  
4     1950-06-18  http://en.wikipedia.org/wiki/1950_Belgian_Gran...  
...          ...                                                ...  
1013  2019-10-13  https://en.wikipedia.org/wiki/2019_Japanese_Gr...  
1014  2019-10-27  https://en.wikipedia.org/wiki/2019_Mexican_Gra...  
1015  2019-11-03  https://en.wikipedia.org/wiki/2019_United_Stat...  
1016  2019-11-17  https://en.wikipedia.org/wiki/2019_Brazilian_G...  
1017  2019-12-01  https://en.wikipedia.org/wiki/2019_Abu_Dhabi_G...  

[1018 rows x 8 columns]

In [77]:
#Loading the race_years csv to use years to pull data from API
race_years = pd.read_csv('years.csv')

In [78]:
#Review the race_years data frame to confirm data for year and rounds is present
race_years

season  round
0       1950      1
1       1950      2
2       1950      3
3       1950      4
4       1950      5
...      ...    ...
1030    2020     13
1031    2020     14
1032    2020     15
1033    2020     16
1034    2020     17

[1035 rows x 2 columns]

In [79]:
#Build rounds for year and rounds in that year

rounds = []
for year in np.array(race_years.season.unique()):
    rounds.append([year, list(race_years[race_years.season == year]['round'])])


In [80]:
#Check what the rounds look like
rounds[:10]

[[1950, [1, 2, 3, 4, 5, 6, 7]],
 [1951, [1, 2, 3, 4, 5, 6, 7, 8]],
 [1952, [1, 2, 3, 4, 5, 6, 7, 8]],
 [1953, [1, 2, 3, 4, 5, 6, 7, 8, 9]],
 [1954, [1, 2, 3, 4, 5, 6, 7, 8, 9]],
 [1955, [1, 2, 3, 4, 5, 6, 7]],
 [1956, [1, 2, 3, 4, 5, 6, 7, 8]],
 [1957, [1, 2, 3, 4, 5, 6, 7, 8]],
 [1958, [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]],
 [1959, [1, 2, 3, 4, 5, 6, 7, 8, 9]]]

In [98]:
#Pull all the results from results dictionary within RaceTable
#Results dictionary with empty lists for data
results = {'season': [],
          'round':[],
           'circuit_id':[],
          'driver': [],
           'date_of_birth': [],
           'nationality': [],
          'constructor': [],
          'grid': [],
          'time': [],
          'status': [],
          'points': [],
          'podium': [],
          'url': []}

#Loop for through the rounds
for n in list(range(len(rounds))):
    for i in rounds[n][1]:
    
        url = 'http://ergast.com/api/f1/{}/{}/results.json' #url to call results
        r = requests.get(url.format(rounds[n][0], i)) #API request to ergast for results
        json = r.json() #Read the returning JSON

#Loop through the json to pull out the results.
        for item in json['MRData']['RaceTable']['Races'][0]['Results']: #Move through JSON to results
            results['season'].append(int(json['MRData']['RaceTable']['Races'][0]['season']))
            results['round'].append(int(json['MRData']['RaceTable']['Races'][0]['round']))
            results['circuit_id'].append(json['MRData']['RaceTable']['Races'][0]['Circuit']['circuitId'])
            results['driver'].append(item['Driver']['driverId'])
            results['date_of_birth'].append(item['Driver']['dateOfBirth'])
            results['nationality'].append(item['Driver']['nationality'])
            results['constructor'].append(item['Constructor']['constructorId'])
            results['grid'].append(int(item['grid']))
            try: #Due to missing times this assist with no returning values. Replace with NaN
                results['time'].append(int(item['Time']['millis']))
            except:#End of try block no value places NaN
                results['time'].append(None)
            results['status'].append(item['status'])
            try: #Try block due to missing points this assist with no returning values. Replace with NaN
                results['points'].append(int(item['points']))
            except:#End of try block no value places NaN
                results['points'].append(None)
            try: #Try block Due to missing podium finishers this assist no returning values. Replaces with NaN
                results['podium'].append(int(item['position']))
            except:#End of try block no value places NaN
                results['podium'].append(None)
            results['url'].append(json['MRData']['RaceTable']['Races'][0]['url'])

#Convert my dictionary to a dataframe
df_results = pd.DataFrame(results)
#Size of my dataframe
print(df_results.shape)

(24507, 13)


In [99]:
#Check to make sure it is a dataframe
type(df_results)

pandas.core.frame.DataFrame

In [100]:
#View the dataframe to confirm
df_results

season  round   circuit_id             driver date_of_birth  \
0        1950      1  silverstone             farina    1906-10-30   
1        1950      1  silverstone            fagioli    1898-06-09   
2        1950      1  silverstone        reg_parnell    1911-07-02   
3        1950      1  silverstone          cabantous    1904-10-08   
4        1950      1  silverstone             rosier    1905-11-05   
...       ...    ...          ...                ...           ...   
24502    2020     17   yas_marina         giovinazzi    1993-12-14   
24503    2020     17   yas_marina             latifi    1995-06-29   
24504    2020     17   yas_marina    kevin_magnussen    1992-10-05   
24505    2020     17   yas_marina  pietro_fittipaldi    1996-06-25   
24506    2020     17   yas_marina              perez    1990-01-26   

      nationality   constructor  grid       time        status  points  \
0         Italian          alfa     1  8003600.0      Finished     9.0   
1         Italian          alfa     2  8006200.0      Finished     6.0   
2         British          alfa     4  8055600.0      Finished     4.0   
3          French          lago     6        NaN       +2 Laps     3.0   
4          French          lago     9        NaN       +2 Laps     2.0   
...           ...           ...   ...        ...           ...     ...   
24502     Italian          alfa    14        NaN        +1 Lap     0.0   
24503    Canadian      williams    18        NaN        +1 Lap     0.0   
24504      Danish          haas    20        NaN        +1 Lap     0.0   
24505   Brazilian          haas    17        NaN       +2 Laps     0.0   
24506     Mexican  racing_point    19        NaN  Transmission     0.0   

       podium                                                url  
0           1  http://en.wikipedia.org/wiki/1950_British_Gran...  
1           2  http://en.wikipedia.org/wiki/1950_British_Gran...  
2           3  http://en.wikipedia.org/wiki/1950_British_Gran...  
3           4  http://en.wikipedia.org/wiki/1950_British_Gran...  
4           5  http://en.wikipedia.org/wiki/1950_British_Gran...  
...       ...                                                ...  
24502      16  https://en.wikipedia.org/wiki/2020_Abu_Dhabi_G...  
24503      17  https://en.wikipedia.org/wiki/2020_Abu_Dhabi_G...  
24504      18  https://en.wikipedia.org/wiki/2020_Abu_Dhabi_G...  
24505      19  https://en.wikipedia.org/wiki/2020_Abu_Dhabi_G...  
24506      20  https://en.wikipedia.org/wiki/2020_Abu_Dhabi_G...  

[24507 rows x 13 columns]

In [101]:
#Check for missing values
for c in df_results.columns:
    miss=df_results[c].isnull().sum()
    if miss > 0:
        print('{} has {} missing values'.format(c,miss))
    else:
        print('{} has NO missing values'.format(c))

season has NO missing values
round has NO missing values
circuit_id has NO missing values
driver has NO missing values
date_of_birth has NO missing values
nationality has NO missing values
constructor has NO missing values
grid has NO missing values
time has 17916 missing values
status has NO missing values
points has 39 missing values
podium has NO missing values
url has NO missing values


Note: Missing data from times. Of the 24507 rows there are 17916 missing. This is expected if car or drivers did not finish that race. So they should not be removed.

In [103]:
#View columns in dataframe
df_results.columns

Index(['season', 'round', 'circuit_id', 'driver', 'date_of_birth',
       'nationality', 'constructor', 'grid', 'time', 'status', 'points',
       'podium', 'url'],
      dtype='object')

In [104]:
#Remove not needed columns
df_results.drop(columns=['url','date_of_birth','nationality'],axis=1, inplace = True)


In [105]:
#Review after removing columns
df_results.columns

Index(['season', 'round', 'circuit_id', 'driver', 'constructor', 'grid',
       'time', 'status', 'points', 'podium'],
      dtype='object')

In [108]:
#Update the columns names
df_results.rename(columns={'circuit_id':'circuit_ref','driver':'driver_ref','grid':'starting_position',
                           'points':'points_earned','podium':'finishing_position'}, inplace = True)

In [111]:
#Confirm changes to columns names
df_results.columns

Index(['season', 'round', 'circuit_ref', 'driver_ref', 'constructor',
       'starting_position', 'time', 'status', 'points_earned',
       'finishing_position'],
      dtype='object')

In [115]:
#Calculate the minutes in new column called time_in_min
df_results['time_in_min'] = df_results.time/60000


In [117]:
#Preview the df confirm changes
df_results.head()

season  round  circuit_ref   driver_ref constructor  starting_position  \
0    1950      1  silverstone       farina        alfa                  1   
1    1950      1  silverstone      fagioli        alfa                  2   
2    1950      1  silverstone  reg_parnell        alfa                  4   
3    1950      1  silverstone    cabantous        lago                  6   
4    1950      1  silverstone       rosier        lago                  9   

        time    status  points_earned  finishing_position  time_in_min  
0  8003600.0  Finished            9.0                   1   133.393333  
1  8006200.0  Finished            6.0                   2   133.436667  
2  8055600.0  Finished            4.0                   3   134.260000  
3        NaN   +2 Laps            3.0                   4          NaN  
4        NaN   +2 Laps            2.0                   5          NaN

In [118]:
#Drop time column as it is not needed
df_results.drop(columns=['time'],axis=1, inplace = True)

In [119]:
#Confirm success
df_results.head()

season  round  circuit_ref   driver_ref constructor  starting_position  \
0    1950      1  silverstone       farina        alfa                  1   
1    1950      1  silverstone      fagioli        alfa                  2   
2    1950      1  silverstone  reg_parnell        alfa                  4   
3    1950      1  silverstone    cabantous        lago                  6   
4    1950      1  silverstone       rosier        lago                  9   

     status  points_earned  finishing_position  time_in_min  
0  Finished            9.0                   1   133.393333  
1  Finished            6.0                   2   133.436667  
2  Finished            4.0                   3   134.260000  
3   +2 Laps            3.0                   4          NaN  
4   +2 Laps            2.0                   5          NaN

In [120]:
#Review the columns again
df_results.columns

Index(['season', 'round', 'circuit_ref', 'driver_ref', 'constructor',
       'starting_position', 'status', 'points_earned', 'finishing_position',
       'time_in_min'],
      dtype='object')

In [122]:
#Rearrange column placement
df_results= df_results[['season', 'round', 'circuit_ref', 'driver_ref', 'constructor',
       'starting_position', 'time_in_min','status', 'points_earned', 'finishing_position']]

In [124]:
#Confirm changes
df_results.head()

season  round  circuit_ref   driver_ref constructor  starting_position  \
0    1950      1  silverstone       farina        alfa                  1   
1    1950      1  silverstone      fagioli        alfa                  2   
2    1950      1  silverstone  reg_parnell        alfa                  4   
3    1950      1  silverstone    cabantous        lago                  6   
4    1950      1  silverstone       rosier        lago                  9   

   time_in_min    status  points_earned  finishing_position  
0   133.393333  Finished            9.0                   1  
1   133.436667  Finished            6.0                   2  
2   134.260000  Finished            4.0                   3  
3          NaN   +2 Laps            3.0                   4  
4          NaN   +2 Laps            2.0                   5

In [125]:
df_results.shape

(24507, 10)

Note:Above in Mileston 2 (full version has it)there is a df that has circuit id. I need that to add to the results df to allow me to connect it to other tables in Milestone 5. I am making that merger here.

In [131]:
#Preview df_circuit to confirm data I need is present
df_circuit_id.head()

circuit_id   circuitRef                            name  circuit_city  \
0           1  albert_park  Albert Park Grand Prix Circuit     Melbourne   
1           2       sepang    Sepang International Circuit  Kuala Lumpur   
2           3      bahrain   Bahrain International Circuit        Sakhir   
3           4    catalunya  Circuit de Barcelona-Catalunya      Montmeló   
4           5     istanbul                   Istanbul Park      Istanbul   

  circuit_country  
0       Australia  
1        Malaysia  
2         Bahrain  
3           Spain  
4          Turkey

In [132]:
#Rename the columns to make the merger happen
df_circuit_id.rename(columns={'circuitRef':'circuit_ref'}, inplace = True)

In [133]:
#Confirm changes took
df_circuit_id.head()

circuit_id  circuit_ref                            name  circuit_city  \
0           1  albert_park  Albert Park Grand Prix Circuit     Melbourne   
1           2       sepang    Sepang International Circuit  Kuala Lumpur   
2           3      bahrain   Bahrain International Circuit        Sakhir   
3           4    catalunya  Circuit de Barcelona-Catalunya      Montmeló   
4           5     istanbul                   Istanbul Park      Istanbul   

  circuit_country  
0       Australia  
1        Malaysia  
2         Bahrain  
3           Spain  
4          Turkey

In [158]:
#Join dataframe to add a circuit id to help connect other data
df_results_final = pd.merge(df_results,df_circuit_id,on=['circuit_ref'],how='outer')

In [159]:
#Review dataframe
df_results_final

season  round    circuit_ref   driver_ref   constructor  \
0      1950.0    1.0    silverstone       farina          alfa   
1      1950.0    1.0    silverstone      fagioli          alfa   
2      1950.0    1.0    silverstone  reg_parnell          alfa   
3      1950.0    1.0    silverstone    cabantous          lago   
4      1950.0    1.0    silverstone       rosier          lago   
...       ...    ...            ...          ...           ...   
24504  2020.0   12.0       portimao       latifi      williams   
24505  2020.0   12.0       portimao        kvyat    alphatauri   
24506  2020.0   12.0       portimao       stroll  racing_point   
24507     NaN    NaN  port_imperial          NaN           NaN   
24508     NaN    NaN          hanoi          NaN           NaN   

       starting_position  time_in_min            status  points_earned  \
0                    1.0   133.393333          Finished            9.0   
1                    2.0   133.436667          Finished            6.0   
2                    4.0   134.260000          Finished            4.0   
3                    6.0          NaN           +2 Laps            3.0   
4                    9.0          NaN           +2 Laps            2.0   
...                  ...          ...               ...            ...   
24504               20.0          NaN           +2 Laps            0.0   
24505               13.0          NaN           +2 Laps            0.0   
24506               12.0          NaN  Collision damage            0.0   
24507                NaN          NaN               NaN            NaN   
24508                NaN          NaN               NaN            NaN   

       finishing_position  circuit_id                                name  \
0                     1.0           9                 Silverstone Circuit   
1                     2.0           9                 Silverstone Circuit   
2                     3.0           9                 Silverstone Circuit   
3                     4.0           9                 Silverstone Circuit   
4                     5.0           9                 Silverstone Circuit   
...                   ...         ...                                 ...   
24504                18.0          75  Autódromo Internacional do Algarve   
24505                19.0          75  Autódromo Internacional do Algarve   
24506                20.0          75  Autódromo Internacional do Algarve   
24507                 NaN          72        Port Imperial Street Circuit   
24508                 NaN          74                Hanoi Street Circuit   

      circuit_city circuit_country  
0      Silverstone              UK  
1      Silverstone              UK  
2      Silverstone              UK  
3      Silverstone              UK  
4      Silverstone              UK  
...            ...             ...  
24504     Portimão        Portugal  
24505     Portimão        Portugal  
24506     Portimão        Portugal  
24507   New Jersey             USA  
24508        Hanoi         Vietnam  

[24509 rows x 14 columns]

In [160]:
#Check for missing values
for c in df_results_final.columns:
    miss=df_results_final[c].isnull().sum()
    if miss > 0:
        print('{} has {} missing values'.format(c,miss))
    else:
        print('{} has NO missing values'.format(c))

season has 2 missing values
round has 2 missing values
circuit_ref has NO missing values
driver_ref has 2 missing values
constructor has 2 missing values
starting_position has 2 missing values
time_in_min has 17918 missing values
status has 2 missing values
points_earned has 41 missing values
finishing_position has 2 missing values
circuit_id has NO missing values
name has NO missing values
circuit_city has NO missing values
circuit_country has NO missing values


In [161]:
#Keep all rows but rows 24507 and 24508 as they have NaN as season. All results have a season.
df_results_final=df_results_final[df_results_final['season'].notna()]

In [162]:
#Check for missing values
for c in df_results_final.columns:
    miss=df_results_final[c].isnull().sum()
    if miss > 0:
        print('{} has {} missing values'.format(c,miss))
    else:
        print('{} has NO missing values'.format(c))

season has NO missing values
round has NO missing values
circuit_ref has NO missing values
driver_ref has NO missing values
constructor has NO missing values
starting_position has NO missing values
time_in_min has 17916 missing values
status has NO missing values
points_earned has 39 missing values
finishing_position has NO missing values
circuit_id has NO missing values
name has NO missing values
circuit_city has NO missing values
circuit_country has NO missing values


In [144]:
#Review dataframe
df_results_final.head()

season  round  circuit_ref   driver_ref constructor  starting_position  \
0  1950.0    1.0  silverstone       farina        alfa                1.0   
1  1950.0    1.0  silverstone      fagioli        alfa                2.0   
2  1950.0    1.0  silverstone  reg_parnell        alfa                4.0   
3  1950.0    1.0  silverstone    cabantous        lago                6.0   
4  1950.0    1.0  silverstone       rosier        lago                9.0   

   time_in_min    status  points_earned  finishing_position  circuit_id  \
0   133.393333  Finished            9.0                 1.0           9   
1   133.436667  Finished            6.0                 2.0           9   
2   134.260000  Finished            4.0                 3.0           9   
3          NaN   +2 Laps            3.0                 4.0           9   
4          NaN   +2 Laps            2.0                 5.0           9   

                  name circuit_city circuit_country  
0  Silverstone Circuit  Silverstone              UK  
1  Silverstone Circuit  Silverstone              UK  
2  Silverstone Circuit  Silverstone              UK  
3  Silverstone Circuit  Silverstone              UK  
4  Silverstone Circuit  Silverstone              UK

In [166]:
#Check df types as they have changed
df_results_final.dtypes

season                float64
round                 float64
circuit_ref            object
driver_ref             object
constructor            object
starting_position     float64
time_in_min           float64
status                 object
points_earned         float64
finishing_position    float64
circuit_id              int64
name                   object
circuit_city           object
circuit_country        object
dtype: object

In [167]:
#Update data types
df_results_final['season'] = df_results_final['season'].astype(int)
df_results_final['round'] = df_results_final['round'].astype(int)
df_results_final['starting_position'] = df_results_final['starting_position'].astype(int)
df_results_final['finishing_position'] = df_results_final['finishing_position'].astype(int)


<ipython-input-167-4abcb333d50e>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_results_final['season'] = df_results_final['season'].astype(int)
<ipython-input-167-4abcb333d50e>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_results_final['round'] = df_results_final['round'].astype(int)
<ipython-input-167-4abcb333d50e>:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the docume

In [168]:
#Preview the df
df_results_final.head()

season  round  circuit_ref   driver_ref constructor  starting_position  \
0    1950      1  silverstone       farina        alfa                  1   
1    1950      1  silverstone      fagioli        alfa                  2   
2    1950      1  silverstone  reg_parnell        alfa                  4   
3    1950      1  silverstone    cabantous        lago                  6   
4    1950      1  silverstone       rosier        lago                  9   

   time_in_min    status  points_earned  finishing_position  circuit_id  \
0   133.393333  Finished            9.0                   1           9   
1   133.436667  Finished            6.0                   2           9   
2   134.260000  Finished            4.0                   3           9   
3          NaN   +2 Laps            3.0                   4           9   
4          NaN   +2 Laps            2.0                   5           9   

                  name circuit_city circuit_country  
0  Silverstone Circuit  Silverstone              UK  
1  Silverstone Circuit  Silverstone              UK  
2  Silverstone Circuit  Silverstone              UK  
3  Silverstone Circuit  Silverstone              UK  
4  Silverstone Circuit  Silverstone              UK

Note: I can't change the points_earned field since it has NaN. Pandas will not allow non-numeric data types to be made numeric. I am stuck with float.